core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -1197,27 +1205,43 @@ wasi_sock_recv(wasm_exec_env_t exec_env, wasi_fd_t sock, iovec_app_t *ri_data,
         || !validate_native_addr(ri_data, (uint32)total_size))
         return __WASI_EINVAL;
 
-    /* recv ri_data one by one */
+    /* receive and scatter*/
+    for (i = 0; i < ri_data_len; i++, ri_data++) {
+        total_size += ri_data->buf_len;
+    }
+    if (total_size >= UINT32_MAX
+        || !(buf_begin = wasm_runtime_malloc((uint32)total_size))) {
+        return __WASI_ENOMEM;
+    }
+
     *ro_data_len = 0;
-    for (i = 0; i < ri_data_len; ri_data++, i++) {
-        void *buf;
-        size_t bytes_recv;
+    err = wasmtime_ssp_sock_recv(curfds, sock, buf_begin, total_size,
+                                 (size_t *)ro_data_len);
+    if (err != __WASI_ESUCCESS) {
+        goto fail;
+    }
 
+    buf = buf_begin;
+    ri_data = ri_data_orig;
+    for (i = 0; i < ri_data_len; ri_data++, i++) {
+        char *native_addr;
         if (!validate_app_addr(ri_data->buf_offset, ri_data->buf_len)) {
-            return __WASI_EINVAL;
+            err = __WASI_EINVAL;
+            goto fail;
         }
 
-        buf = (void *)addr_app_to_native(ri_data->buf_offset);
-        err = wasmtime_ssp_sock_recv(curfds, sock, buf, ri_data->buf_len,
-                                     &bytes_recv);
-        if (err != __WASI_ESUCCESS) {
-            return err;
-        }
-        *ro_data_len += bytes_recv;
+        native_addr = (void *)addr_app_to_native(ri_data->buf_offset);
+        bh_memcpy_s(native_addr, ri_data->buf_len, buf, ri_data->buf_len);
+        buf += ri_data->buf_len;
```

Comment:
Shall we check whether `(size_t)(buf - buf_begin) >= *ro_data_len`? If yes, break and don't copy again:
```C
if ((size_t)(buf - buf_begin) >= *ro_data_len)
    break;
```

Reply:
Because indicated the length of the buffer is `total_size`, the received data length should be less than `total_size`. Or else, `recv` will write out of the buffer.
```
    err = wasmtime_ssp_sock_recv(curfds, sock, buf_begin, total_size,
                                 (size_t *)ro_data_len);
```